File: scripts/build_screen_icon_references.mjs
import {createServer} from 'node:http';
import {mkdir, readFile, writeFile} from 'node:fs/promises';
import path from 'node:path';
import {createRequire} from 'node:module';

const {chromium} = createRequire(import.meta.url)('playwright');
const root = path.resolve(import.meta.dirname, '..');
const fixtures = path.join(root, 'tests/fixtures/recognition');
const output = path.join(root, 'dist/recognition-screen-icons');
const manifest = JSON.parse(await readFile(path.join(fixtures, 'manifest.json'), 'utf8'));
const types = new Map([['.mjs','text/javascript'],['.json','application/json'],['.jpg','image/jpeg'],['.png','image/png'],['.webp','image/webp'],['.html','text/html']]);

const server = createServer(async (request, response) => {
  try {
    const relative = decodeURIComponent(new URL(request.url, 'http://localhost').pathname).replace(/^\/+/, '');
    const file = path.resolve(root, relative || 'dist/index.html');
    if (!file.startsWith(root + path.sep)) throw Error('outside root');
    response.setHeader('content-type', types.get(path.extname(file)) || 'application/octet-stream');
    response.end(await readFile(file));
  } catch { response.statusCode = 404; response.end(); }
});

await mkdir(output, {recursive: true});
await new Promise(resolve => server.listen(0, '127.0.0.1', resolve));
const browser = await chromium.launch({channel:'msedge', headless:true});
try {
  const page = await browser.newPage();
  const origin = `http://127.0.0.1:${server.address().port}`;
  await page.goto(`${origin}/dist/index.html`);
  const records = [];
  for (const [filename, expected] of Object.entries(manifest.lead)) {
    const fixtureData = await page.evaluate(async ({origin, filename}) => {
      const recognition = await import(`${origin}/dist/local-image-recognition.mjs`);
      const image = new Image(); image.src = `${origin}/tests/fixtures/recognition/${filename}`; await image.decode();
      const result = [];
      for (const side of ['mine','opp']) {
        const rects = recognition.detectedLeadRects(image, side);
        for (let slot=0; slot<6; slot++) result.push({side,slot,data:recognition.cropRecognitionRect(image,rects[slot]).toDataURL('image/webp',.95)});
      }
      return {crops: result, signature: recognition.imageSignature(image)};
    }, {origin, filename});
    for (const image of fixtureData.crops) {
      const name = expected[image.side][image.slot];
      if (typeof name !== 'string') continue;
      const target = `lead-${String(records.length).padStart(3,'0')}.webp`;
      await writeFile(path.join(output,target),Buffer.from(image.data.split(',')[1],'base64'));
      records.push({name,src:`recognition-screen-icons/${target}`,kind:'lead',side:image.side,slot:image.slot,fixture:filename,fixtureSignature:fixtureData.signature});
    }
  }
  const partyNames = manifest.party.names;
  for (const filename of ['party-stats.jpg','party-ability.jpg']) {
    const fixtureData = await page.evaluate(async ({origin,filename}) => {
      const recognition=await import(`${origin}/dist/local-image-recognition.mjs`);
      const image=new Image();image.src=`${origin}/tests/fixtures/recognition/${filename}`;await image.decode();
      return {signature:recognition.imageSignature(image),crops:recognition.fixedSlotRects('party',image.width,image.height).map((rect,slot)=>({slot,data:recognition.cropRecognitionRect(image,rect).toDataURL('image/webp',.95)}))};
    },{origin,filename});
    for(const image of fixtureData.crops){const target=`party-${filename.includes('stats')?'stats':'ability'}-${image.slot}.webp`;await writeFile(path.join(output,target),Buffer.from(image.data.split(',')[1],'base64'));records.push({name:partyNames[image.slot],src:`recognition-screen-icons/${target}`,kind:'party',slot:image.slot,fixture:filename,fixtureSignature:fixtureData.signature});}
  }
  await writeFile(path.join(root,'dist/recognition-screen-map.json'),JSON.stringify(records,null,2)+'\n');
  console.log(`Built ${records.length} exact runtime icon references.`);
} finally { await browser.close(); server.close(); }

File: scripts/build_screen_icon_references.py
"""Build small, non-sensitive icon references from accepted screenshots."""

import json
from pathlib import Path

from PIL import Image


ROOT = Path(__file__).resolve().parents[1]
FIXTURES = ROOT / "tests" / "fixtures" / "recognition"
OUTPUT = ROOT / "dist" / "recognition-screen-icons"

PARTY_CENTERS = [(.119, .280), (.535, .280), (.119, .488), (.535, .488), (.119, .690), (.535, .690)]


def crop(image, center_x, center_y, size):
    width, height = image.size
    pixels = size * height
    box = (
        round(center_x * width - pixels / 2),
        round(center_y * height - pixels / 2),
        round(center_x * width + pixels / 2),
        round(center_y * height + pixels / 2),
    )
    return image.crop(box).resize((96, 96), Image.Resampling.LANCZOS)


def main():
    OUTPUT.mkdir(parents=True, exist_ok=True)
    manifest = json.loads((FIXTURES / "manifest.json").read_text(encoding="utf-8"))
    references = []
    for filename, sides in manifest["lead"].items():
        image = Image.open(FIXTURES / filename).convert("RGB")
        compact = image.width / image.height < 2.2
        first_y, step_y, size = (.176, .113, .145) if compact else (.202, .119, .135)
        for side, center_x in (("mine", .322), ("opp", .810)):
            for index, name in enumerate(sides[side]):
                if not isinstance(name, str):
                    continue
                target = f"lead-{len(references):02d}.webp"
                crop(image, center_x, first_y + step_y * index, size).save(OUTPUT / target, "WEBP", quality=95)
                references.append({"name": name, "src": f"recognition-screen-icons/{target}", "kind": "lead", "side": side, "fixture": filename})

    for filename in ("party-stats.jpg", "party-ability.jpg"):
        image = Image.open(FIXTURES / filename).convert("RGB")
        for index, (name, (center_x, center_y)) in enumerate(zip(manifest["party"]["names"], PARTY_CENTERS)):
            target = f"party-{filename[6:11]}-{index}.webp"
            crop(image, center_x, center_y, .09).save(OUTPUT / target, "WEBP", quality=95)
            references.append({"name": name, "src": f"recognition-screen-icons/{target}", "kind": "party", "fixture": filename})

    (ROOT / "dist" / "recognition-screen-map.json").write_text(
        json.dumps(references, ensure_ascii=False, indent=2) + "\n",
        encoding="utf-8",
    )


if __name__ == "__main__":
    main()
